virtcontainers/sandbox.go

```diff
@@ -931,6 +934,122 @@ func (s *Sandbox) removeNetwork() error {
 	return s.network.remove(s, s.networkNS, s.networkNS.NetNsCreated)
 }
 
+// UpdateRoute updates the sandbox route table (e.g. for portmapping support).
+func (s *Sandbox) UpdateRoute(routes []*grpc.Route) error {
+	if err := s.agent.updateRoutes(routes); err != nil {
+		return err
+	}
+	s.networkNS.Routes = routes
+	return s.storage.storeSandboxNetwork(s.id, s.networkNS)
+}
+
+func (s *Sandbox) generateNetInfo(inf grpc.Interface) (NetworkInfo, error) {
+	hw, err := net.ParseMAC(inf.HwAddr)
+	if err != nil {
+		return NetworkInfo{}, err
+	}
+
+	var addrs []netlink.Addr
+	for _, addr := range inf.IPAddresses {
+		netlinkAddrStr := fmt.Sprintf("%s/%s", addr.Address, addr.Mask)
+		netlinkAddr, err := netlink.ParseAddr(netlinkAddrStr)
+		if err != nil {
+			return NetworkInfo{}, fmt.Errorf("could not parse %q: %v", netlinkAddrStr, err)
+		}
+
+		addrs = append(addrs, *netlinkAddr)
+	}
+
+	return NetworkInfo{
+		Iface: NetlinkIface{
+			LinkAttrs: netlink.LinkAttrs{
+				Name:         inf.Name,
+				HardwareAddr: hw,
+				MTU:          int(inf.Mtu),
+			},
+			Type: "",
+		},
+		Addrs: addrs,
+	}, nil
+}
+
+func (s *Sandbox) generateInf(netInfo NetworkInfo) (grpc.Interface, error) {
+	iface := netInfo.Iface
+	hw := iface.HardwareAddr.String()
+
+	return grpc.Interface{
+		Device:      iface.Name,
+		Name:        iface.Name,
+		IPAddresses: generateIPAddr(netInfo.Addrs, s.Logger()),
+		Mtu:         uint64(iface.MTU),
+		HwAddr:      hw,
+	}, nil
+}
+
+// AddNetwork adds new nic to the sandbox.
+func (s *Sandbox) AddNetwork(inf grpc.Interface) error {
+	netInfo, err := s.generateNetInfo(inf)
+	if err != nil {
+		return err
+	}
+
+	endpoint, err := createVirtualNetworkEndpoint(len(s.networkNS.Endpoints), netInfo.Iface.Name, s.config.NetworkConfig.InterworkingModel)
+	if err != nil {
+		return err
+	}
+	endpoint.SetProperties(netInfo)
+	if err := endpoint.HotAttach(s.hypervisor); err != nil {
+		return err
+	}
+
+	// Update the sandbox storage
+	s.networkNS.Endpoints = append(s.networkNS.Endpoints, endpoint)
+	if err := s.storage.storeSandboxNetwork(s.id, s.networkNS); err != nil {
+		return err
+	}
+
+	// Add network for vm
+	if err := s.agent.updateNetwork(endpoint); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ListNetwork lists all nics and their configurations in the sandbox.
+func (s *Sandbox) ListNetwork() ([]grpc.Interface, []*grpc.Route, error) {
+	return s.networkNS.Interfaces, s.networkNS.Routes, nil
+}
+
+// UpdateNetwork updates the configuration of an existing network interface.
+func (s *Sandbox) UpdateNetwork(interfaces []grpc.Interface) error {
+	if err := s.deleteNetworks(); err != nil {
```

Comment:
The semantics of `UpdateNetwork` is a bit weird. It will unplug all existing NICs and re-plug them together with any additional NICs. E.g, if a VM has one NIC and we want to add another one. Why do we have to remove the existing one first? I am not sure that is the best way of doing network hotplug.

Can we have `AddNetwork` and `DeleteNetwork` at least since that would cover the most common use cases and does not wipe out all existing NICs during the process?

Reply:
Deleting existing ones is definitely not good,  we can have a better implementation.

Reply:
@WeiZhang555 @bergwolf I agree with this, we don't want to remove an existing interface since the point of `UpdateNetwork()` is to update something that already exists.

Reply:
OK. I will rewrite it